openproblems/tasks/spatial_decomposition/methods/cell2location.py

```diff
@@ -0,0 +1,112 @@
+from ....tools.decorators import method
+from ....tools.utils import check_version
+from ..utils import split_sc_and_sp
+
+import numpy as np
+
+
+@method(
+    method_name="Cell2location",
+    paper_name="Cell2location maps fine-grained cell types in spatial transcriptomics",
+    paper_url="https://www.nature.com/articles/s41587-021-01139-4",
+    paper_year=2022,
+    code_url="https://github.com/BayraktarLab/cell2location",
+    image="openproblems-python-extras",
+)
+def stereoscope(adata, test=False, max_epochs_sc=None, max_epochs_sp=None):
+
+    from cell2location.models import Cell2location
+    from cell2location.models import RegressionModel
+
+    if test:
+        max_epochs_sp = max_epochs_sp or 10
+        max_epochs_sc = max_epochs_sc or 10
+        num_samples = 10
+    else:  # pragma: nocover
+        max_epochs_sc = max_epochs_sc or 250
+        max_epochs_sp = max_epochs_sp or 30000
+        num_samples = 1000
+
+    adata_sc, adata = split_sc_and_sp(adata)
+
+    if "tech" in adata_sc.obs.columns:
```

Comment:
I don't think this column is used in other methods -- we should either document it or make it unavailable. @LuckyMD @giovp thoughts?

Reply:
Cell2location benefits from the batch label for 10X reaction/sample/batch for both single cell and spatial data. The single-cell regression model can also use the technology information as an extra categorical covariate (could be useful for the pancreas data, right?). Does the dataset for this task have the batch obs column?

Reply:
Do you mean adding these additional columns, technology and batch to https://github.com/openproblems-bio/openproblems/blob/main/openproblems/tasks/spatial_decomposition/api.py for this task?

Is the scRNA reference for this task composed of multiple batches and technologies?

Reply:
If multiple technologies, this dataset is suboptimal for simulating spatial data because it introduces unusual technical differences between spots.

In our paper, we simulated data using data from one technology with minor batch effects to minimise weird results.

Reply:
In this case, the pancreas data is composed of multiple batches/technologies. The question is whether or not we want to include this in the task API -- I lean towards not because currently the other methods are not written to use this information.

Reply:
> If multiple technologies, this dataset is suboptimal for simulating spatial data because it introduces unusual technical differences between spots.
> 
> In our paper, we simulated data using data from one technology with minor batch effects to minimise weird results.

You raise an important point. Perhaps we want to add another dataset which doesn't have this to evaluate how well methods do with and without this technical difficulty

Reply:
TBH, I don't know if the cell2location reference NB regression model can be applied to multiple batches and technologies without specifying those batches and technologies. So I added hardcoded average computation as a default and an additional dataset which uses the NB regression model.

Hard-coded reference signature option is anyway necessary for SS2 scRNA data.

Reply:
I'm out of my depth here. @almaan @giovp @LuckyMD any thoughts on the above?

Reply:
Maybe we can continue this conversation in a new issue about improving datasets?

Reply:
Maybe we can continue this conversation in a new issue about improving the realism of datasets for this task?

Reply:
@vitkl can you open an issue with a summary of this discussion and tag me? Thanks